src/main/java/com/google/sps/servlets/ProfileServlet.java

```diff
@@ -0,0 +1,118 @@
+import javax.servlet.http.HttpServletResponse;
+import org.json.simple.JSONObject;
+
+/** Servlet that posts and gets user profiles in Datastore */
+@WebServlet("/profile")
+public class ProfileServlet extends HttpServlet {
+  
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    response.setContentType("application/json");
+    JSONObject responseMap = new JSONObject();
+    
+    if (userService.isUserLoggedIn()) {
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      Query query = new Query("Profile")
+        .setFilter(new Query.FilterPredicate("id", Query.FilterOperator.EQUAL, userService.getCurrentUser().getUserId()));
+      PreparedQuery results = datastore.prepare(query);
+      Entity entity = results.asSingleEntity();
+      if (entity == null) {
+        responseMap.put("hasProfile", false);
+      } else {
+        long id = (long) entity.getProperty("id");
+        String userName = (String) entity.getProperty("userName");
+        boolean vegetarian = (boolean) entity.getProperty("vegetarian");
+        boolean vegan = (boolean) entity.getProperty("vegan");
+        boolean glutenFree = (boolean) entity.getProperty("glutenFree");
+        boolean dairyFree = (boolean) entity.getProperty("dairyFree");
+        String[] allergies = (String[]) entity.getProperty("allergies");
+
+        Profile profileObject = new Profile(id, userName, vegetarian, vegan, glutenFree, dairyFree, allergies);
+        responseMap.put("profile", profileObject);
+        responseMap.put("hasProfile", true);
+      }
+
+    } else {
+      String errorMessage = "User needs to log in to see profile.";
+      responseMap.put("error", errorMessage);
+    }
+
+    Gson gson = new Gson();   
+    String json = gson.toJson(responseMap);
+    response.setContentType("application");
+    response.getWriter().println(json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    response.setContentType("application/json");
+    JSONObject responseMap = new JSONObject();
+
+    if (!userService.isUserLoggedIn()) {
+      String errorMessage = "User needs to log in to change profile.";
+      responseMap.put("error", errorMessage);
+      
+    } else {
+      String id = userService.getCurrentUser().getUserId();
+      String userName = request.getParameter("userName");
+      boolean vegetarian = Boolean.parseBoolean(request.getParameter("vegetarian"));
+      boolean vegan = Boolean.parseBoolean(request.getParameter("vegan"));
+      boolean glutenFree = Boolean.parseBoolean(request.getParameter("glutenFree"));
+      boolean dairyFree = Boolean.parseBoolean(request.getParameter("dairyFree"));
+      String[] allergies = request.getParameterValues("allergies");
```

Comment:
Do you want to set an upper limit for the array size?

Reply:
I could set an upper limit but in case a person has a ton of allergies I don't want to assume that there is an upper bound on the amount of allergies a person can have and remove the extra ones.